c2corg_api/models/document.py

```diff
@@ -53,15 +67,101 @@ class (e.g. waypoint, route, ...) a separate table will be created, which
     # TODO constraint that there is at least one locale
     locales = relationship('DocumentLocale')
 
-    _ATTRIBUTES = ['document_id', 'protected', 'redirects_to', 'quality']
+    __mapper_args__ = {
+            'version_id_col': _DocumentMixin.version,
+            'version_id_generator': lambda version: uuid.uuid4().hex
+    }
+
+    _ATTRIBUTES_WHITELISTED = \
+        ['document_id', 'version']
 
-    def to_archive(self, doc):
+    _ATTRIBUTES = \
+        _ATTRIBUTES_WHITELISTED + ['protected', 'redirects_to', 'quality']
+
+    @abc.abstractmethod
+    def to_archive(self):
+        """Create an `Archive*` instance with the same attributes.
+        This method is supposed to be implemented by child classes.
+        """
+        return
+
+    def _to_archive(self, doc):
+        """Copy the attributes of this document into a passed in
+        `Archive*` instance.
+        """
         copy_attributes(self, doc, Document._ATTRIBUTES)
         return doc
 
     def get_archive_locales(self):
         return [locale.to_archive() for locale in self.locales]
 
+    def update(self, other):
+        """Copies the attributes from `other` to this document.
+        Also updates all locales.
+        """
+        copy_attributes(other, self, Document._ATTRIBUTES_WHITELISTED)
+
+        for locale_in in other.locales:
+            locale = self.get_locale(locale_in.culture)
+            if locale:
+                locale.update(locale_in)
+                locale.document_id = self.document_id
+            else:
+                self.locales.append(locale_in)
+
+    def get_versions(self):
+        """Get the version hashs of this document and of all its locales.
+        """
+        version = self.version
+
+        locale_versions = {}
+        for locale in self.locales:
+            locale_versions[locale.culture] = locale.version
+
+        return {
+            'document': version,
+            'locales': locale_versions
+        }
+
+    def get_update_type(self, old_versions):
+        """Get the update type (only figures have changed, only locales have
+        changed, both have changed or nothing) and the languages that have
+        changed.
+        This is done by comparing the old version hashs (before flushing to
```

Comment:
In the tests below, `get_update_type` is called after the flush.
Is the `merge` operation where SQLAlchemy will update the version?